lib/mongoid/fields.rb

```diff
@@ -757,38 +779,72 @@ def generated_methods
         post_processed_defaults.delete_one(name)
       end
 
+      # Create a field for the given name and options.
+      #
+      # @param [ Symbol ] name The name of the field.
+      # @param [ Hash ] options The hash of options.
+      #
+      # @return [ Field ] The created field.
+      #
+      # @api private
       def field_for(name, options)
         opts = options.merge(klass: self)
-        type_mapping = TYPE_MAPPINGS[options[:type]]
-        opts[:type] = type_mapping || unmapped_type(options)
-        if !opts[:type].is_a?(Class)
-          raise Errors::InvalidFieldType.new(self, name, options[:type])
+        opts[:type] = retrieve_and_validate_type(name, options[:type])
+        return Fields::Localized.new(name, opts) if options[:localize]
+        return Fields::ForeignKey.new(name, opts) if options[:identity]
+        Fields::Standard.new(name, opts)
+      end
+
+       # Get the class for the given type.
+       #
+       # @param [ Symbol ] name The name of the field.
+       # @param [ Symbol | Class ] type The type of the field.
+       #
+       # @return [ Class ] The type of the field.
+       #
+       # @raises [ Mongoid::Errors::InvalidFieldType ] if given an invalid field
+       #   type.
+       #
+       # @api private
+       def retrieve_and_validate_type(name, type)
+        type_mapping = TYPE_MAPPINGS[type]
+        result = type_mapping || unmapped_type(type)
+        if !result.is_a?(Class)
+          raise Errors::InvalidFieldType.new(self, name, type)
         else
-          if INVALID_BSON_CLASSES.include?(opts[:type])
-            warn_message = "Using #{opts[:type]} as the field type is not supported. "
-            if opts[:type] == BSON::Decimal128
+          if INVALID_BSON_CLASSES.include?(result)
+            warn_message = "Using #{result} as the field type is not supported. "
+            if result == BSON::Decimal128
               warn_message += "In BSON <= 4, the BSON::Decimal128 type will work as expected for both storing and querying, but will return a BigDecimal on query in BSON 5+."
             else
               warn_message += "Saving values of this type to the database will work as expected, however, querying them will return a value of the native Ruby Integer type."
             end
             Mongoid.logger.warn(warn_message)
           end
         end
-        return Fields::Localized.new(name, opts) if options[:localize]
-        return Fields::ForeignKey.new(name, opts) if options[:identity]
-        Fields::Standard.new(name, opts)
+        result
       end
 
-      def unmapped_type(options)
-        if "Boolean" == options[:type].to_s
+      # Returns the type of the field if the type was not in the TYPE_MAPPINGS
+      # hash.
+      #
+      # @param [ Symbol | Class ] type The type of the field.
+      #
+      # @return [ Class ] The type of the field.
+      #
+      # @api private
+      def unmapped_type(type)
+        if "Boolean" == type.to_s
```

Comment:
I would like to get rid of this special case at some point but it's not in scope of the this PR.

Reply:
Im actually going to expand this method in the typed array PR lol